### PR TITLE
When possible, use Reference6 to read ProjectReference metadata such as `ReferenceOutputAssembly`

### LINKED ISSUE #11163: UIDelay: nuget.packagemanagement.visualstudio.dll!NuGet.PackageManagement.VisualStudio.VsCoreProjectSystemReferenceReader+<GetProjectReferencesAsync>d__ - managed projects
During mixed project restores, this UI delay can happen. 

[1257734](
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1257734/?triage=true)
[1426094](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1426094)

When managed projects are involved, we are doing a very inefficient walk in https://github.com/NuGet/NuGet.Client/blob/c1a0597e5498bcf3bf0938b3dd0170edbf0b2d20/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs#L54. 

This issue tracks rewriting that. 

cc @davkean

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11163

Regression? Last working version:

## Description

This is an attempt to the fix the UI delay caused by the *double* walk of the references..

I setup a test repo with the following relevant projects.

1. CPlusPlus.vcxproj - [CpsProjectSystemReferencesReader](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/CpsProjectSystemReferenceReader.cs)
1. LegacyMSbuildProj.msbuildproj [VSCoreProjectSystemReferencesReader](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs)
1. LegacyPackagesConfig.csproj [VSCoreProjectSystemReferencesReader](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs)
1. ServiceFabric.sfproj [VSCoreProjectSystemReferencesReader](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs)
1. Stateless.csproj [VsManagedLanguagesProjectSystemServices](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs)

The project references are as follows:

LegacyPackagesConfig.csproj
    => CPlusPlus.vcxproj
    => LegacyMSbuildProj.msbuildproj
    => Stateless.csproj
ServiceFabric.sfproj
    => Stateless.csproj
LegacyMSbuildProj.msbuildproj
    => Stateless.csproj

When walking each of the 3 above projects for ProjectReference, the following happens: 

* All the `LegacyPackagesConfig` project references can be cast to Reference6. 
* None of the `ServiceFabric` project references can be cast to Reference6, but can be cast to Reference3.
* None of the `LegacyMSbuildProj` project references can be cast to Reference6, but can be cast to Reference3.

The proposal is the following:

Currently the behavior is:

* walk the tree for projects that should be excluded.
* walk the tree for all projects and skip the ones that should be excluded. 

The first part is the expensive part. 

The proposal is: 

* Walk the tree for all projects and if the projects can be cast to `Reference6`, then just use the metadata from there, otherwise use `Reference3` and keep track of the fact that we couldn't get the metadata.
* If *any* of the above projects couldn't be cast to `Reference6`, walk the tree for the excluded projects the expensive way to *avoid* any functional regressions (which are frankly, pretty unlikely, but we can't break the experience.)

This should fix the UI delay for *most* project types.

@lifengl @davkean
Based on the above findings, we'd need CPS to implement Reference6 - https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1431168

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [x] Test exception  - Existing behavior. 
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs [hunk 1]
```diff
@@ -40,6 +43,10 @@ internal class VsCoreProjectSystemReferenceReader
 
             _vsProjectAdapter = vsProjectAdapter;
             _threadingService = threadingService;
+
+            _referenceMetadata = Array.CreateInstance(typeof(string), 1);
```
Comment (davkean) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: This can be just `new string [] { "ReferenceOutputAssembly" }`, that's exactly what `Array.CreateInstance(typeof(string), 1);` is creating, just in a roundabout way.

**Review by davkean — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: Fixed.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs [hunk 2]
```diff
@@ -81,17 +82,30 @@ internal class VsCoreProjectSystemReferenceReader
...
                         var childProjectPath = reference3.SourceProject.GetFullProjectPath();
 
+
                         // Skip projects which have ReferenceOutputAssembly=false
-                        if (!string.IsNullOrEmpty(childProjectPath)
-                            && !excludedProjects.Contains(childProjectPath, StringComparer.OrdinalIgnoreCase))
+                        var reference6 = childReference as Reference6;
+                        var addProject = true;
+
+                        if (reference6 != null)
```
Comment (loic-sharma) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: You can delete line 87 with this:

```suggestion
                        if (childReference is Reference6 reference6)
```

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs [hunk 3]
```diff
@@ -81,17 +82,30 @@ internal class VsCoreProjectSystemReferenceReader
 
                         var childProjectPath = reference3.SourceProject.GetFullProjectPath();
 
+
```
Comment (loic-sharma) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: FYI there's an extra line here

**Comment by nkolev92** [comment]
@NuGet/nuget-client 

Can I please get a review here?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs [hunk 4]
```diff
@@ -124,7 +136,26 @@ internal class VsCoreProjectSystemReferenceReader
...
+                {
+                    results = results.Where(e => excludedProjects.Contains(e.ProjectPath, StringComparer.OrdinalIgnoreCase)).ToList();
+                }
+            }
+
             return results;
+
+            static string GetReferenceMetadataValue(Array metadataElements, Array metadataValues)
+            {
+                if (metadataElements == null || metadataValues == null || metadataValues.Length == 0)
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: Should we add `metadataElements.Length == 0` condition also?

Another question that I have is, `metadataElements` argument value is not used except for null or empty checks. Do we need this parameter?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs [hunk 5]
```diff
@@ -124,7 +136,26 @@ internal class VsCoreProjectSystemReferenceReader
                 logger.LogVerbose(message);
             }
 
+            if (hasProjectsWithUnresolvedMetadata)
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs:139]: I am new to this code path and trying to understand it. Should we set `hasProjectsWithUnresolvedMetadata = true` in the exception handler just after setting `hasMissingReferences = true`. https://github.com/NuGet/NuGet.Client/blob/6522b99940ba336c5ae97a5820930063f01ffea6/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs#L106

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs [hunk 6]
```diff
@@ -124,7 +136,26 @@ internal class VsCoreProjectSystemReferenceReader
...
+            {
+                IList<string> excludedProjects = await GetExcludedProjectsAsync(logger);
+                if (excludedProjects.Count > 0)
+                {
+                    results = results.Where(e => excludedProjects.Contains(e.ProjectPath, StringComparer.OrdinalIgnoreCase)).ToList();
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: Callers to `GetProjectReferencesAsync` method are invoking `.ToList()` method explicitly (couple of examples are listed below). May be invoking `.ToList()` here is not required. It is possible I am missing something here as I am looking at this code path for the first time. 

For e.g. https://github.com/NuGet/NuGet.Client/blob/d2442f173798ae9f2d95cda34cc156b4517afcdf/src/NuGet.Core/NuGet.PackageManagement/Projects/MSBuildNuGetProject.cs#L687-L688

https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectJsonNuGetProject.cs#L223-L224

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs:139]: I don't think so.

They have a different purpose. 
hasMissingReferences  is about a warning to the customer telling them that some refs weren't resolved. 
Whether the project-system implements ref3 or ref6 is not super important. 

hasProjectsWithUnresolvedMetadata basically tells us whether we saw any projects that we couldn't cast to Ref6.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs (hunk 6 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: I get your point, but I'm trying to focus on the implementation of the current method here. 

I think it'd be better if the calling methods would call `AsList` instead, but ideally that's done in a different PR.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: I think we can just remove metadataElements.

**Review by zivkan — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs [hunk 7]
```diff
@@ -207,7 +249,7 @@ private IEnumerable<Reference> GetVSProjectReferences()
             return excludedReferences;
         }
 
-        private bool IsProjectReference(Reference3 reference, Common.ILogger logger)
+        private bool IsProjectReference(Reference reference, Common.ILogger logger)
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: Why did you change this one? It's still passing same `Reference3` type pointer here.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: I think it was mostly the process than the result that led to this changes. 
It was just something when I was refactoring, coming up with cleanest way of writing the code. 

I reverted it.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs (hunk 6 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: Isn't it `results = results.Where(e => !excludedProjects.Contains(e.ProjectPath, StringComparer.OrdinalIgnoreCase)).ToList();` based on context? Exclude mean we don't want, right? I'm trying to grasp why only project in excluded are now returned, correct me if I'm wrong.

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: If above is the case maybe below is more concise. It looks `Except` uses [HashSet internally](http://index/?leftProject=System.Core&leftSymbol=ktweftexyual&file=System%5CLinq%5CEnumerable.cs&line=899) could be more performant on large list, but could be more overhead on if 2nd list is small.
image

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: I think the perf concerns are not super important cause `ReferenceOutputAssembly` is not frequently used, and those list are always small. 

I'll take a look at the condition.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs]: Fixed. 
There isn't really a test case that can actually repro that reliably.

ServiceFabric projects are supposed to be it, but the refs are not resolved there.

**Review by erdembayar — APPROVED** [review]
